### PR TITLE
feat: proactive decision and milestone detection (#182)

### LINKED ISSUE #182: feat(memory): add proactive decision and milestone detection with save suggestions
## Problem

Memory is only saved at session-stop or explicitly by the user. Important decisions and milestones during a session can be lost if the session is compacted, or if the user forgets to save. The memory-strategy.md (Phase D) calls for heuristic pattern detection that identifies save-worthy events during conversation and injects non-blocking save suggestions into Claude's context.

## Proposed Solution

Create a `ProactiveDetector` service and integrate it into `PromptSubmitHandler` for non-blocking save suggestions.

| File | Action | Details |
|------|--------|---------|
| `src/lib/domain/interfaces/IProactiveDetector.ts` | Create | Define `IProactiveDetection` and `IProactiveDetector` interfaces |
| `src/lib/infrastructure/services/ProactiveDetector.ts` | Create | Factory `createProactiveDetector()` with decision confirmation and milestone detection |
| `src/lib/domain/interfaces/index.ts` | Modify | Export `IProactiveDetector`, `IProactiveDetection` |
| `src/lib/infrastructure/services/index.ts` | Modify | Export `createProactiveDetector` |
| `src/lib/infrastructure/di/tokens.ts` | Modify | Add `ProactiveDetector` symbol to `INFRA_TOKENS` |
| `src/lib/infrastructure/di/bootstrap.ts` | Modify | Register `ProactiveDetector`; inject into `PromptSubmitHandler` |
| `src/lib/application/handlers/PromptSubmitHandler.ts` | Modify | Add optional `IProactiveDetector` parameter; append save suggestion to `additionalContext` when detected |
| `src/lib/application/mediator/requests/PromptSubmitRequest.ts` | Modify | Add optional `previousAssistantMessage` field |
| `src/lib/infrastructure/adapters/claude/user-prompt-submit.ts` | Modify | Extract and pass `previous_assistant_message` from hook input |
| `tests/unit/.../ProactiveDetector.test.ts` | Create | ~18 test cases for decision and milestone detection |
| `tests/unit/.../PromptSubmitHandler.test.ts` | Modify | ~6 new test cases for proactive suggestion integration |

### Decision detection

- User prompt matches confirmation pattern (`yes/ok/sounds good/let's go with/do that/go ahead/approved/agreed`)
- AND `previousAssistantMessage` contains option-presenting language (`option|approach|alternative|should we|we could|either|or|recommend|suggest`)
- Returns `{ shouldSuggest: true, suggestedFact: "DECISION: <topic>", factType: 'decision', confidence: 'high' }`

### Milestone detection

- PR creation: `created PR|opened PR|pull request #\d+`
- Test success: `all tests pass|tests passing|0 fail`
- Completion: `done|finished|completed|shipped|merged`
- Version bump: `bumped version|v\d+\.\d+`
- Returns `{ shouldSuggest: true, suggestedFact: "MILESTONE: <text>", factType: 'milestone', confidence: 'medium' }`

### Suggestion format (non-blocking)

```
[Memory suggestion: This looks like a <decision/milestone>. Consider saving: "<suggestedFact>". Use /memory to save.]
```

## Acceptance Criteria

- [ ] `ProactiveDetector` detects decision confirmations when preceded by option-presenting assistant messages
- [ ] `ProactiveDetector` detects milestones: PR creation, test success, completion, version bump
- [ ] Suggestions are non-blocking (injected as context text, not auto-saved)
- [ ] `PromptSubmitRequest` accepts optional `previousAssistantMessage`
- [ ] Hook adapter passes previous assistant message when available
- [ ] All existing tests pass without modification
- [ ] Registered in DI container and injected into PromptSubmitHandler
- [ ] New code covered by unit tests (~24 test cases)
- [ ] All tests pass (`npm test`)
- [ ] Linting passes (`npm run lint`)
- [ ] TypeScript compiles (`npx tsc --noEmit`)

## References

- Part of #178
- Depends on PR C (#181)
- `.dev/features/memory/memory-strategy.md` (Phase D: Proactive Memory Hooks, Section 1)
- `.dev/features/memory/memory-strategy-execution-plan.md` (Issue 4 / PR D)
- `src/lib/application/handlers/PromptSubmitHandler.ts`
- `src/lib/application/mediator/requests/PromptSubmitRequest.ts`
- `src/lib/infrastructure/adapters/claude/user-prompt-submit.ts`

### PR DESCRIPTION
**Pull request by TonyCasey**
## Summary

- Add `ProactiveDetector` service that detects save-worthy events (decisions, milestones) during conversation
- Decision detection: user confirmation after assistant presents options (requires both user prompt + assistant context)
- Milestone detection: PR creation, test success, completion keywords, version bumps
- Non-blocking: suggestions injected as `[Memory suggestion: ...]` context text, no auto-save
- `PromptSubmitRequest` extended with optional `previousAssistantMessage` for proactive detection context
- Hook adapter passes `previous_assistant_message` from hook input

## Test plan

- [x] 25 new ProactiveDetector tests (decision detection, milestone detection, priority, edge cases)
- [x] 8 new PromptSubmitHandler proactive integration tests
- [x] All 1489 existing tests pass
- [x] Lint clean
- [x] Build succeeds

Closes #182

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Proactive decision & milestone detection that non-blockingly suggests memory saves during conversations, using prior assistant context to improve suggestions
  * Confidence-weighted retrieval ranking (6-tier) with improved prioritization and display

* **Tests**
  * Expanded unit test coverage for proactive detection and prompt-submit integration, covering decisions, milestones, priority behavior, and edge cases

### CODE REVIEW

**Comment by TonyCasey** [comment]
Addressed CodeRabbit review: fixed empty decision topics in \ by trimming the assistant message first and falling back to 'Decision made' when no sentence is found (commit 87a6749).

File: src/lib/infrastructure/services/ProactiveDetector.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
+/**
+ * ProactiveDetector
+ *
+ * Detects save-worthy events during conversation and returns
+ * non-blocking save suggestions. Looks for:
+ * - Decision confirmations (user confirms after assistant presents options)
+ * - Milestones (PR creation, test success, completion, version bump)
+ */
+
+import type { IProactiveDetector, IProactiveDetection } from '../../domain/interfaces/IProactiveDetector';
```
[the remark this answers is not in the document]
Comment (TonyCasey) [reply src/lib/infrastructure/services/ProactiveDetector.ts]: Switched the import to use the interfaces index export (IProactiveDetector/IProactiveDetection).
